src/PluginCreator.ts

```diff
@@ -0,0 +1,193 @@
+import * as ts from 'typescript'
+import * as resolve from 'resolve'
+import compareVersions from 'compare-versions'
+
+export type TransformerPlugin = {
+    before?: ts.TransformerFactory<ts.SourceFile>
+    after?: ts.TransformerFactory<ts.SourceFile>
+    afterDeclaration?: ts.TransformerFactory<ts.SourceFile>
+} | ts.TransformerFactory<ts.SourceFile>
+
+export type FactoryType = 'ls' | 'program' | 'opts' | 'checker'
+
+export type PluginConfig = {
+    transform: string
+    type?: FactoryType
+    after?: boolean
+    before?: boolean
+    afterDeclaration?: boolean
+    [options: string]: any;
+} | (ts.PluginImport & {
+    [options: string]: any
+})
+
```

Comment:
Can you add `name?: string` to main interface and remove `[options: string]: any;` cause your interface just transforms into just `any`
After that you will see couple of bugs below :)

Reply:
Can't use pattern matching here. 
This not works: ``` {name: void, transform: string} | {name: string} ```
Changed to single type.